File: src/domains/environment/rbac.test.ts
import { describe, expect, it } from "vitest";
import { canReadRole, canWriteRole } from "./rbac.js";

describe("RBAC helpers", () => {
  it("viewer can read not write", () => {
    expect(canReadRole("viewer")).toBe(true);
    expect(canWriteRole("viewer")).toBe(false);
  });
  it("editor can read and write", () => {
    expect(canReadRole("editor")).toBe(true);
    expect(canWriteRole("editor")).toBe(true);
  });
  it("undefined denies", () => {
    expect(canReadRole(undefined)).toBe(false);
    expect(canWriteRole(undefined)).toBe(false);
  });
});

File: internal/services/rbac_test.go
package services

import (
	"testing"

	"flora-hive/internal/domain/models"
)

func TestCanReadRole(t *testing.T) {
	t.Parallel()
	if !CanReadRole(models.RoleViewer) {
		t.Fatal("viewer should read")
	}
	if !CanReadRole(models.RoleEditor) {
		t.Fatal("editor should read")
	}
	if CanReadRole("") {
		t.Fatal("empty role should not read")
	}
	if CanReadRole(models.MemberRole("admin")) {
		t.Fatal("unknown role should not read")
	}
}

func TestCanWriteRole(t *testing.T) {
	t.Parallel()
	if CanWriteRole(models.RoleViewer) {
		t.Fatal("viewer should not write")
	}
	if !CanWriteRole(models.RoleEditor) {
		t.Fatal("editor should write")
	}
	if CanWriteRole("") {
		t.Fatal("empty role should not write")
	}
	if CanWriteRole(models.MemberRole("admin")) {
		t.Fatal("unknown role should not write")
	}
}
